contracts/cw20-base/src/contract.rs

```diff
@@ -128,14 +131,25 @@ pub fn instantiate(
     };
     TOKEN_INFO.save(deps.storage, &data)?;
 
+    let logo_string = &mut String::from("");
     if let Some(marketing) = msg.marketing {
         let logo = if let Some(logo) = marketing.logo {
-            verify_logo(&logo)?;
+            let logo_type = verify_logo(&logo)?;
             LOGO.save(deps.storage, &logo)?;
 
             match logo {
-                Logo::Url(url) => Some(LogoInfo::Url(url)),
-                Logo::Embedded(_) => Some(LogoInfo::Embedded),
+                Logo::Url(url) => {
+                    logo_string.push_str(url.as_str());
+                    Some(LogoInfo::Url(url))
+                },
+                Logo::Embedded(binary_type) => {
+                    let b64_logo = match binary_type {
+                        EmbeddedLogo::Svg(binary) => binary.to_base64(),
+                        EmbeddedLogo::Png(binary) => binary.to_base64(),
+                    };
+                    logo_string.push_str((logo_type + ":" + b64_logo.as_str()).as_str());
```

Comment:
5KB limited so I think it's fine;

Reply:
Doesn't this basically mean those 5KB are stored on the chain twice? Once in contract storage, second time in the events log.

So you have an added gas cost and extra storage used. If you just perform a contract query after the tx, you have none of that.

Reply:
Ah yes you're right, it doubled gas cost, I removed in event to save some gas consumption;
I think only allowing url is better since it's more efficient, few bytes compare to large logo binary files, just an idea

Reply:
Well. What problem are we solving here? I would assume it's not a huge problem if an offchain service has to query for the logo of a currency here and there, and I would assume an offchain service would cache logos for various currencies.

There are events with `"action", "upload_logo"` that the offchain service can track to know when to update its cache.